pubtools/_pulp/tasks/delete.py

```diff
@@ -0,0 +1,709 @@
+    def get_modules(self, module_names):
+        criteria = self.unit_criteria(
+            ModulemdUnit, self._module_search_criteria(module_names)
+        )
+        mods_f = self.search_content(criteria)
+        mods_f = f_map(
+            mods_f,
+            partial(self.log_missing_units, unit_names=module_names, unit_attr="nsvca"),
+        )
+        return mods_f
+
+    def unit_criteria(self, unit_type, partial_crit):
+        criteria = Criteria.and_(
+            Criteria.with_unit_type(unit_type), Criteria.or_(*partial_crit)
+        )
+        return criteria
+
+    def _module_search_criteria(self, module_names):
+        part_crit = []
+
+        for mod_name in module_names:
+            part_crit.append(self._module_criteria(mod_name))
+
+        return part_crit
+
+    def _module_criteria(self, module_name):
+        crit = []
+        mod_dict = self._get_nsvca(module_name)
+        for m_part, value in mod_dict.items():
+            crit.append(Criteria.with_field(m_part, value))
+        return Criteria.and_(*crit)
+
+    def _get_nsvca(self, module_name):
+        mod_parts = ["name", "stream", "version", "context", "arch"]
+        nsvca = module_name.split(":", 4)
+        mod_dict = {}
+        for i, p in enumerate(nsvca):
+            if mod_parts[i] == "version":
+                mod_dict[mod_parts[i]] = int(p)
+            else:
+                mod_dict[mod_parts[i]] = p
+
+        return mod_dict
+
+    def _file_search_criteria(self, file_names):
+        part_crit = []
+
+        for file_name in file_names:
+            part_crit.append(self._file_criteria(file_name))
+
+        return part_crit
+
+    def _file_criteria(self, file_name):
+        return Criteria.with_field("path", file_name)
+
+    def _rpm_search_criteria(self, rpm_names, signing_keys):
+        part_crit = []
+        signing_keys = [s.lower() for s in signing_keys] if signing_keys else [None]
+
+        for rpm_name in rpm_names:
+            part_crit.append(self._rpm_criteria(rpm_name, signing_keys))
+
+        return part_crit
+
+    def _rpm_criteria(self, filename, signing_keys):
+        return Criteria.and_(
+            Criteria.with_field("filename", filename),
+            Criteria.with_field("signing_key", Matcher.in_(signing_keys)),
+        )
+
+    def search_content(self, criteria):
+        return self.pulp_client.search_content(criteria=criteria)
+
+    def log_missing_units(self, searched_units, unit_attr, unit_names):
+        found = []
+
+        for unit in searched_units:
+            found.append(getattr(unit, unit_attr))
+
+        missing = set(unit_names) - set(found)
+        if missing:
+            missing = sorted(list(missing))
+            LOG.warning("Requested unit(s) doesn't exist: %s", ", ".join(missing))
+
+        LOG.info("%s unit(s) found for deletion", len(found))
+        # TODO: collect missing units "NOT FOUND"
+
+        return searched_units
+
+    @step("Mapping units to repo")
```

Comment:
I don't think this should be a step, there's nothing asynchronous here at all, won't it always complete in less than a second?

Reply:
Ack

Reply:
Removed it as there's nothing asynchronous.